File: backend/generators/mcg.py
"""
Multiplicative Congruential Generator (MCG)

Formula: X(n+1) = (a * X(n)) mod m

Parameters:
    seed (X0): initial value (must be odd and non-zero when m is a power of 2)
    a: multiplier
    m: modulus
    count: number of values to generate
"""


def generate(seed: int, a: int, m: int, count: int) -> dict:
    """
    Generate pseudo-random numbers using the Multiplicative Congruential Generator.

    Returns a dict with:
      - sequence: list of generated integers
      - steps: list of dicts with intermediate calculations per iteration
      - normalized: list of floats in [0, 1)
    """
    if m <= 0:
        raise ValueError("Modulus m must be a positive integer greater than 0.")
    if seed == 0:
        raise ValueError("Seed (X0) must be non-zero for the MCG.")
    if count <= 0:
        raise ValueError("Count must be a positive integer.")
    if count > 10000:
        raise ValueError("Count must not exceed 10,000.")

    sequence = []
    steps = []
    x = seed % m
    if x == 0:
        raise ValueError("Seed mod m must be non-zero for the MCG.")

    for i in range(count):
        x_prev = x
        x_next = (a * x_prev) % m
        normalized = x_next / m if m > 0 else 0.0
        steps.append({
            "iteration": i + 1,
            "x_prev": x_prev,
            "formula": f"({a} × {x_prev}) mod {m}",
            "x_next": x_next,
            "normalized": round(normalized, 6),
        })
        sequence.append(x_next)
        x = x_next

    return {
        "sequence": sequence,
        "steps": steps,
        "normalized": [round(v / m, 6) for v in sequence],
    }

File: frontend-threejs/src/lib/generators/mcg.ts
/**
 * Multiplicative Congruential Generator (MCG)
 * Formula: X(n+1) = (a * X(n)) mod m
 */

import type { GeneratorResult, MCGParams, StepDetail } from './types';

export function generate(params: MCGParams): GeneratorResult {
  const { seed, a, m, count } = params;

  if (m <= 0) throw new Error('El módulo m debe ser un entero positivo mayor que 0.');
  if (seed === 0) throw new Error('La semilla (X0) debe ser distinta de cero para el MCG.');
  if (count <= 0) throw new Error('La cantidad debe ser un entero positivo.');
  if (count > 10000) throw new Error('La cantidad no debe exceder 10,000.');

  let x = ((seed % m) + m) % m;
  if (x === 0) throw new Error('La semilla mod m debe ser distinta de cero para el MCG.');

  const sequence: number[] = [];
  const steps: StepDetail[] = [];

  for (let i = 0; i < count; i++) {
    const xPrev = x;
    const xNext = (a * xPrev) % m;
    const normalized = m > 0 ? xNext / m : 0;

    steps.push({
      iteration: i + 1,
      xPrev,
      formula: `(${a} × ${xPrev}) mod ${m}`,
      xNext,
      normalized: Math.round(normalized * 1000000) / 1000000,
    });

    sequence.push(xNext);
    x = xNext;
  }

  return {
    sequence,
    steps,
    normalized: sequence.map((v) => Math.round((v / m) * 1000000) / 1000000),
  };
}
